shell/platform/darwin/macos/framework/Source/FlutterKeyEmbedderHandler.mm

```diff
@@ -0,0 +1,605 @@
+  _sendEvent(flutterEvent, nullptr, nullptr);
+}
+
+- (void)sendModifierEventOfType:(BOOL)shouldDown
+                      timestamp:(NSTimeInterval)timestamp
+                        keyCode:(unsigned short)keyCode
+                       callback:(FlutterKeyHandlerCallback)callback {
+  uint64_t physicalKey = GetPhysicalKeyForKeyCode(keyCode);
+  uint64_t logicalKey = GetLogicalKeyForModifier(keyCode, physicalKey);
+  if (physicalKey == 0 || logicalKey == 0) {
+    NSLog(@"Unrecognized modifier key: keyCode 0x%hx, physical key 0x%llx", keyCode, physicalKey);
+    callback(TRUE);
+    return;
+  }
+  FlutterKeyEvent flutterEvent = {
+      .struct_size = sizeof(FlutterKeyEvent),
+      .timestamp = GetFlutterTimestampFrom(timestamp),
+      .type = shouldDown ? kFlutterKeyEventTypeDown : kFlutterKeyEventTypeUp,
+      .physical = physicalKey,
+      .logical = logicalKey,
+      .character = nil,
+      .synthesized = callback == nil,
+  };
+  [self updateKey:physicalKey asPressed:shouldDown ? logicalKey : 0];
+  if (callback != nil) {
+    [self sendPrimaryFlutterEvent:flutterEvent callback:callback];
+  } else {
+    _sendEvent(flutterEvent, nullptr, nullptr);
+  }
+}
+
+- (void)handleDownEvent:(NSEvent*)event callback:(FlutterKeyHandlerCallback)callback {
+  uint64_t physicalKey = GetPhysicalKeyForKeyCode(event.keyCode);
+  uint64_t logicalKey = GetLogicalKeyForEvent(event, physicalKey);
+  [self synchronizeModifiers:event.modifierFlags ignoringFlags:0 timestamp:event.timestamp];
+
+  bool isARepeat = event.isARepeat;
+  NSNumber* pressedLogicalKey = _pressingRecords[@(physicalKey)];
+  if (pressedLogicalKey != nil && !isARepeat) {
+    callback(TRUE);
+    return;
+  }
+  bool isSynthesized = false;
+
+  if (pressedLogicalKey == nil) {
+    [self updateKey:physicalKey asPressed:logicalKey];
+  }
+
+  FlutterKeyEvent flutterEvent = {
+      .struct_size = sizeof(FlutterKeyEvent),
+      .timestamp = GetFlutterTimestampFrom(event.timestamp),
+      .type = isARepeat ? kFlutterKeyEventTypeRepeat : kFlutterKeyEventTypeDown,
+      .physical = physicalKey,
+      .logical = pressedLogicalKey == nil ? logicalKey : [pressedLogicalKey unsignedLongLongValue],
+      .character = getEventString(event.characters),
+      .synthesized = isSynthesized,
+  };
+  [self sendPrimaryFlutterEvent:flutterEvent callback:callback];
+}
+
+- (void)handleUpEvent:(NSEvent*)event callback:(FlutterKeyHandlerCallback)callback {
+  NSAssert(!event.isARepeat, @"Unexpected repeated Up event: keyCode %d, char %@, charIM %@",
+           event.keyCode, event.characters, event.charactersIgnoringModifiers);
+  [self synchronizeModifiers:event.modifierFlags ignoringFlags:0 timestamp:event.timestamp];
+
+  uint64_t physicalKey = GetPhysicalKeyForKeyCode(event.keyCode);
+  NSNumber* pressedLogicalKey = _pressingRecords[@(physicalKey)];
+  if (pressedLogicalKey == nil) {
+    NSAssert(FALSE,
+             @"Received key up event that has not been pressed: "
```

Comment:
I was about to say that macOS tracks a key tap (down then up) with the window it starts with therefore this won't happen, but then I realized that this might happen in add-to-app scenarios. I've added this case back.